src/lib/salad/rtree.c

```diff
@@ -1055,6 +1055,33 @@ rtree_remove(struct rtree *tree, const struct rtree_rect *rect, record_t obj)
 	return true;
 }
 
+int
+rtree_replace(struct rtree *tree, struct rtree_rect *old_rect, record_t old_obj,
+	      struct rtree_rect *new_rect, record_t new_obj, bool *removed)
+{
+	/*
+	 * The data structure has no OOM checks, so we have to resreve the
+	 * amount of matras block allocations potentially enough to perform
+	 * the operations required. It's based on the amount of extents we
+	 * currently reserve in the memtx space prior to replace according
+	 * to the smallest page size: 512 (so we sum-up the reserved memory
+	 * and divide it into 512 and got 512 pages).
+	 */
+	if (matras_alloc_reserve(&tree->mtab, 512) != 0)
+		return -1;
```

Comment:
I don't think it's a good idea to refer to the code we're deleting. What's the page size in rtree?

Reply:
512 for 1-dimensional tree, 8192 for 20-dimentional tree.

We reserve 16 extents prior to replace, size of an extent is 16kb, so in total we reserve 16 * 16kb = 262144 bytes.

This is either 512 1-dimentional tree pages or 32 20-dimentional tree pages.

So I took the higher value, now we have

Reply:
AFAIK the most common tree is 2-dimensional, not 1-dimensional.

I took a quick look at the rtree source code and found the constant `RTREE_MAX_HEIGHT` equal to 16. Can you deduce the reservation count from it? Wouldn't `2 * max height` be always enough?

Reply:
[Deferred](https://github.com/tarantool/tarantool/pull/10734#discussion_r1847764459) this part of the patch.